control-plane/agents/src/bin/core/controller/reconciler/node/nexus.rs

```diff
@@ -0,0 +1,173 @@
+use crate::controller::{
+    reconciler::{PollContext, TaskPoller},
+    resources::{
+        operations::ResourcePublishing, operations_helper::OperationSequenceGuard, OperationGuard,
+        ResourceMutex,
+    },
+    task_poller::{PollEvent, PollResult, PollTimer, PollerState},
+};
+use agents::errors::SvcError;
+use common_lib::types::v0::{
+    store::{node::NodeSpec, volume::VolumeSpec},
+    transport::{NodeId, RepublishVolume, VolumeId, VolumeShareProtocol},
+};
+
+/// Node drain reconciler.
+#[derive(Debug)]
+pub(super) struct NodeNexusReconciler {
+    counter: PollTimer,
+}
+impl NodeNexusReconciler {
+    /// Return a new `Self`.
+    pub(super) fn new() -> Self {
+        Self {
+            counter: PollTimer::from(1), // sets the reconciler polling rate
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TaskPoller for NodeNexusReconciler {
+    async fn poll(&mut self, context: &PollContext) -> PollResult {
+        let nodes = context.specs().get_nodes();
+        let mut results = Vec::with_capacity(nodes.len());
+
+        for node in nodes {
+            results.push(check_and_drain_node(context, &node).await);
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+
+    async fn poll_event(&mut self, context: &PollContext) -> bool {
+        matches!(context.event(), PollEvent::TimedRun)
+    }
+}
+
+/// Republish the specified volume on any node other than its current node.
+async fn republish_volume(
+    volume: &mut OperationGuard<ResourceMutex<VolumeSpec>, VolumeSpec>,
+    context: &PollContext,
+    vol_uuid: &VolumeId,
+    frontend_node: &NodeId,
+) -> Result<(), SvcError> {
+    let request = RepublishVolume::new(
+        vol_uuid.clone(),
+        None,
+        frontend_node.clone(),
+        VolumeShareProtocol::Nvmf,
+        false,
+    );
+    tracing::info!(
+        volume.uuid = vol_uuid.as_str(),
+        "Attempting to republish volume"
+    );
+
+    volume.republish(context.registry(), &request).await?;
+
+    Ok(())
+}
+
+/// Drain the specified node if in draining state
+async fn check_and_drain_node(context: &PollContext, node_spec: &NodeSpec) -> PollResult {
+    if !node_spec.is_draining() {
+        return PollResult::Ok(PollerState::Idle);
+    }
+
+    let node_id = node_spec.id();
+    tracing::trace!(node.id = node_spec.id().as_str(), "Draining node");
+    // iterate through all the volumes, find those with a nexus hosted on the node
+    let volumes = context.registry().get_volumes().await;
+
+    let mut volumes_moved = 0;
+    let mut volumes_on_node = 0;
+
+    for volume in volumes.iter() {
+        let vol_spec = volume.spec();
```

Comment:
shouldn't the guard be taken at the start of the loop? Otherwise we still have a race?

Reply:
Now takes the guard at the top of the loop.